applications/nsmrs/pkg/serviceregistryserver/cache.go

```diff
@@ -0,0 +1,151 @@
+func (rc *nseRegistryCache) AddNetworkServiceEndpoint(entry *registry.NSERegistration) (*registry.NSERegistration, error) {
+	rc.Lock()
+	defer rc.Unlock()
+
+	return rc.addNetworkServiceEndpoint(entry)
+}
+
+func (rc *nseRegistryCache) addNetworkServiceEndpoint(entry *registry.NSERegistration) (*registry.NSERegistration, error) {
+	if endpoint, ok := rc.endpoints[entry.NetworkServiceEndpoint.Name]; ok {
+		return nil, errors.Errorf("network service endpoint with name %s already exists: old: %v; new: %v", endpoint.NetworkServiceEndpoint.Name, endpoint, entry)
+	}
+
+	existingEndpoints := rc.GetEndpointsByNs(entry.NetworkService.Name)
+	for _, endpoint := range existingEndpoints {
+		if !proto.Equal(endpoint.NetworkService, entry.NetworkService) {
+			return nil, errors.Errorf("network service already exists with different parameters: old: %v; new: %v", endpoint, entry)
+		}
+	}
+
+	entry.NetworkServiceManager.ExpirationTime = &timestamp.Timestamp{Seconds: time.Now().Add(rc.nseExpirationTimeout).Unix()}
+
+	rc.networkServiceEndpoints[entry.NetworkService.Name] = append(rc.networkServiceEndpoints[entry.NetworkService.Name], entry)
+	rc.endpoints[entry.NetworkServiceEndpoint.Name] = entry
+
+	logrus.Infof("Registered NSE entry %v", entry)
+
+	return entry, nil
+}
+
+func (rc *nseRegistryCache) UpdateNetworkServiceEndpoint(nse *registry.NSERegistration) (*registry.NSERegistration, error) {
+	rc.Lock()
+	defer rc.Unlock()
+
+	if endpoint, ok := rc.endpoints[nse.NetworkServiceEndpoint.Name]; ok {
+		if endpoint.NetworkServiceManager.Name != nse.NetworkServiceManager.Name {
+			return nil, errors.Errorf("network service endpoint with name %s already registered from different NSM: old: %v; new: %v", endpoint.NetworkServiceEndpoint.Name, endpoint, nse)
+		}
+		endpoint.NetworkServiceManager.ExpirationTime = &timestamp.Timestamp{Seconds: time.Now().Add(rc.nseExpirationTimeout).Unix()}
+		return endpoint, nil
+	}
+
+	return rc.addNetworkServiceEndpoint(nse)
+}
+
+// DeleteNetworkServiceEndpoint - remove NSE from cache
+func (rc *nseRegistryCache) DeleteNetworkServiceEndpoint(endpointName string) (*registry.NSERegistration, error) {
+	rc.Lock()
+	defer rc.Unlock()
+
+	delete(rc.endpoints, endpointName)
+	for networkService, endpointList := range rc.networkServiceEndpoints {
+		for i := range endpointList {
+			if endpointList[i].NetworkServiceEndpoint.Name == endpointName {
+				endpoint := endpointList[i]
+				rc.networkServiceEndpoints[networkService] = append(endpointList[:i], endpointList[i+1:]...)
+				return endpoint, nil
+			}
+		}
+	}
+	return nil, errors.Errorf("endpoint %s not found", endpointName)
+}
+
+// GetEndpointsByNs - get Endpoints list from cache by Name
+func (rc *nseRegistryCache) GetEndpointsByNs(networkServiceName string) []*registry.NSERegistration {
+	return rc.networkServiceEndpoints[networkServiceName]
+}
+
+func (rc *nseRegistryCache) StartNSMDTracking() {
+	go func() {
+		for {
+			<-time.After(rc.nseExpirationTimeout / 2)
```

Comment:
I think we need to wait for minimal NSE lifetime left, and react on add/update/remove of NSEs. I do not think we need to wakeup so often.

Reply:
If we will change interval to next NSE timeout each request - it will go through whole NSE list much more often then now